lib/Dialect/Verif/Transforms/PrepareVerifForExportVerilog.cpp

```diff
@@ -0,0 +1,91 @@
+//===- PrepareVerifForExportVerilog.cpp - Formal Preparations --------------*- C++ -*--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Prepare verif dialect for the verillog export.
+//
+//===----------------------------------------------------------------------===//
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Dialect/Verif/VerifPasses.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace circt;
+
+namespace circt {
+namespace verif {
+#define GEN_PASS_DEF_PREPAREVERIFFOREXPORTVERILOG
+#include "circt/Dialect/Verif/Passes.h.inc"
+} // namespace verif
+} // namespace circt
+
+
+using namespace mlir;
+using namespace verif;
+
+namespace {
+struct PrepareVerifForExportVerilogPass
+    : circt::verif::impl::PrepareVerifForExportVerilogBase<
+          PrepareVerifForExportVerilogPass> {
+  void runOnOperation() override;
+};
+
+struct FormalOpConversionPattern : public OpConversionPattern<verif::FormalOp> {
+  using OpConversionPattern<FormalOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(FormalOp op, OpAdaptor operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Create the ports for all the symbolic values
+    SmallVector<hw::PortInfo> ports;
+    for (auto sym : op.getBody().front().getOps<verif::SymbolicValueOp>()) {
+      ports.push_back(
+          hw::PortInfo({{rewriter.getStringAttr("symbolic_value_" +
+                                                std::to_string(ports.size())),
+                         sym.getType(), hw::ModulePort::Input}}));
+    }
+
+    auto moduleOp =
+        rewriter.create<hw::HWModuleOp>(op.getLoc(), op.getNameAttr(), ports);
+
+    // Insert before output op
+    rewriter.setInsertionPointToStart(moduleOp.getBodyBlock());
+
+    // Clone body except for symbolic values which we replace with module
+    // arguments
+    size_t i = 0;
+    for (auto& innerOp : op.getBody().front().getOperations()) {
+      if (dyn_cast<verif::SymbolicValueOp>(innerOp)) {
+        rewriter.replaceAllUsesWith(innerOp.getResult(0), moduleOp.getArgumentForInput(i));
+        i++;
+      } else {
+        rewriter.clone(innerOp);
+      }
+    }
```

Comment:
I wonder if we could avoid cloning the operations here, since we are going to entirely replace the `verif.formal` op with an `hw.module`. For example, you could try to use the `splice()` function on `Block` to take the operations in the `FormalOp` and insert them into the newly-created module. Something like the following might work:
```c++
Block &moduleBlock = moduleOp.getBody().front();
Block &formalBlock = op.getBody().front();
moduleBlock.getOperations().splice(
  moduleBlock.begin(),
  formalBlock.getOperations()
);
```
After that the formal op's body block should be empty. You'd still have to iterate over the body and delete all the symbolic value ops though.

Reply:
I think `splice` should not be used in conversion patterns. These kinds of transformations need to go through the rewriter. I suggest looking at the various inline methods the rewrite provides instead.

Reply:
That's an excellent point. 👍

Reply:
Updated it to use inlineBlockBefore